### PR TITLE
JP-3727: changed psfalign product from quadmodel to cubemodel

### LINKED ISSUE #8746: All frames of the psfalign QuadModel are identical
_Issue [JP-3727](https://jira.stsci.edu/browse/JP-3727) was created on JIRA by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter):_

[This Pull Request](https://github.com/spacetelescope/jwst/pull/8643) implemented a change that simplified the pipeline's coronagraphic data processing such that the alignment of the PSF to the science data is only done on the first science integration, then assumed to be the same for all subsequent integrations.  Prior to this change, the shifted PSF was stored in a 4-dimensional data structure (QuadModel of shape n_sci_integrations, n_psf_integrations, n_rows, n_cols), which made sense when the shift was different on each science integration.  Now, though, this model is identical over the zeroth axis and therefore redundant with itself.  Furthermore, this model uses a lot of memory (see [JP-3724](https://jira.stsci.edu/browse/JP-3724)), and is likely a (the?) cause of the out-of-memory failures reported on multiple datasets in [JWSTDMS-921](https://jira.stsci.edu/browse/JWSTDMS-921) (see comments on that ticket for additional datasets).  The psfalign product should be stored as a CubeModel instead.

### PR DESCRIPTION
**Pull request by emolter**
Resolves [JP-3727](https://jira.stsci.edu/browse/JP-3727)

Closes #8746 

[A previous pull request](https://github.com/spacetelescope/jwst/pull/8643) implemented a change that simplified the pipeline's coronagraphic data processing such that the alignment of the PSF to the science data is only done on the first science integration, then assumed to be the same for all subsequent integrations.  Prior to this change, the shifted PSF was stored in a 4-dimensional data structure (QuadModel of shape n_sci_integrations, n_psf_integrations, n_rows, n_cols), which made sense when the shift was different on each science integration.  Now, though, this model is identical over the zeroth axis and therefore redundant with itself.  Furthermore, this model uses a lot of memory (see [JP-3724](https://jira.stsci.edu/browse/JP-3724)), and is likely a (the?) cause of the out-of-memory failures reported on multiple datasets in [JWSTDMS-921](https://jira.stsci.edu/browse/JWSTDMS-921) (see comments on that ticket for additional datasets).  This PR stores the shifted PSF as a CubeModel instead.

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by emolter** [comment]
[regression tests started here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1683/)

**Comment by emolter** [comment]
Regression tests are all accounted for: 18 of the 22 failures match the ones on the [nightly runs here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST/3018/#showFailuresLink).  The other four (two each for NIRCam and MIRI) are all directly testing the `psfalign` product, which with this PR changed from a QuadModel to a CubeModel, so these failures are expected..  The `_i2d` and other products output by `coron3` are all identical.

File: jwst/coron/tests/test_coron.py
Comment (melanieclarke) [inline jwst/coron/tests/test_coron.py:186]: Remove debug print

**Review by melanieclarke — COMMENTED** [review]
This looks great!  Thanks for the analysis and for putting it together.

File: jwst/coron/tests/test_coron.py
Comment (emolter) [reply jwst/coron/tests/test_coron.py:186]: whoops, thanks for catching that

**Review by melanieclarke — APPROVED** [review]
LGTM, pending concurrence from any of the INS coronography folks.

**Comment by mperrin** [comment]
I am going to defer or hand off doing a review for this one, as I'm not an INS coronagraphy folks anymore. I would suggest that instead @aggle should review and concur for MIRI, and @juliengirard for NIRCam.

**Review by bsunnquist — APPROVED** [review]
Looks good to me - the psf-subtracted images from the PID-1386 test case look reasonable as well.

**Comment by aggle** [comment]
Looks good to me as well!

**Comment by melanieclarke** [comment]
Okay, sounds like we have agreement.  I'll go ahead and merge.  Thanks all!

**Pull request merged**
